addons/beehave/debug/tree_node.gd

```diff
@@ -190,23 +197,59 @@ func _center_nodes_between(left_node: TreeNode, right_node: TreeNode) -> void:
 
 
 func _get_left_contour(node: TreeNode, mod_sum: float, values: Dictionary) -> void:
-	if not values.has(node.y):
-		values[node.y] = node.x + mod_sum
+	var node_left: float = node.x + mod_sum
+	var depth := int(node.y)
+	if not values.has(depth):
+		values[depth] = node_left
 	else:
-		values[node.y] = min(values[node.y], node.x + mod_sum)
+		values[depth] = min(values[depth], node_left)
 
 	mod_sum += node.mod
 	for child in node.children:
 		_get_left_contour(child, mod_sum, values)
 
 
 func _get_right_contour(node: TreeNode, mod_sum: float, values: Dictionary) -> void:
-	var node_right = node.x + mod_sum + node.item.size.x
-	if not values.has(node.y):
-		values[node.y] = node_right
+	var node_right: float = node.x + mod_sum + node.item.layout_size
+	var depth := int(node.y)
+	if not values.has(depth):
+		values[depth] = node_right
 	else:
-		values[node.y] = max(values[node.y], node_right)
+		values[depth] = max(values[depth], node_right)
 
 	mod_sum += node.mod
 	for child in node.children:
 		_get_right_contour(child, mod_sum, values)
+
+
+func _swap_x_y(node: TreeNode) -> void:
+	for child in node.children:
+		_swap_x_y(child)
+
+	var temp := node.x
+	node.x = node.y
+	node.y = temp
+
+
+func _calculate_x(node: TreeNode, offset: int) -> void:
+	node.x = offset
+	var sibling := node.get_most_left_sibling()
+	var max_size: int = node.item.size.x
+	while sibling != null:
+		max_size = max(sibling.item.size.x, max_size)
+		sibling = sibling.get_next_sibling()
+
+	for child in node.children:
+		_calculate_x(child, max_size + offset + LEVEL_DISTANCE * BeehaveUtils.get_editor_scale())
+
+
+func _calculate_y(node: TreeNode, offset: int) -> void:
+	node.y = offset
+	var sibling := node.get_most_left_sibling()
+	var max_size: int = node.item.size.y
+	while sibling != null:
+		max_size = max(sibling.item.size.y, max_size)
+		sibling = sibling.get_next_sibling()
+
+	for child in node.children:
+		_calculate_y(child, max_size + offset + LEVEL_DISTANCE * BeehaveUtils.get_editor_scale())
```

Comment:
I think this could be refactored to one function `_calculate_node_depth`. We could use another property like `layout_size` but reverse (maybe we can call it `node_height` ?) e.g.
```
var node_height: float:
	get:
		return size.x if horizontal else size.y
```
and use it in `var max_size: int = node.item.height` and `max_size = max(sibling.item.height, max_size)`.
Now the only difference between is the first line but we can solve that by assuming a vertical layout (so we use `node.y = offset`) and call `_calculate_node_depth` before we swap x and y, so in `update_positions` something like:
```
	_calculate_node_depth(self, 0)
	if horizontally:
		_swap_x_y(self)
```